File: Leetcode/45_Jump Game2/c++/solution.cpp
/*******************
 * Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Your goal is to reach the last index in the minimum number of jumps.
*******************/
#include <iostream>
#include <fstream>
#include <string>
#include <cstdio>
#include <cmath>
#include <vector>
#include <list>
#include <forward_list>
#include <map>
#include <deque>
#include <algorithm>

using namespace std;

// 每次记录当前能够走到的最远范围，下次迭代直接从最远范围后一个开始搜索。
class Solution {
public:
    int jump(vector<int>& nums) {
        // step: 次数, start: 起点, end: 终点
        int n = nums.size(), step = 0, start = 0, end = 0;
        while (end < n - 1) {
            step++; 
			int maxend = end + 1;
			for (int i = start; i <= end; i++) {
                if (i + nums[i] >= n - 1) return step;
				maxend = max(maxend, i + nums[i]);
			}
            start = end + 1;
            end = maxend;
        }
		return step;
    }
};

int main() {
    return 0;
}

File: Leetcode/45_Jump Game2/python/solution.py
class Solution:
    def jump(self, nums):
        n = len(nums)
        step, start, end = 0, 0, 0

        while end < n-1:
            step += 1
            maxend = end + 1
            for i in range(start, end+1):
                if i + nums[i] >= n - 1:
                    return step;
                maxend = max(maxend, i+nums[i])
            start = end + 1
            end = maxend

        return step
            